### PR TITLE
[NFC] [AutoDiff] Gardening.

### PR DESCRIPTION
**Pull request by dan-zheng**

### CODE REVIEW

**Comment by dan-zheng** [comment]
@swift-ci Please test tensorflow

File: lib/SILOptimizer/Mandatory/Differentiation.cpp [hunk 1]
```diff
@@ -3893,11 +3900,9 @@ class AdjointEmitter final : public SILInstructionVisitor<AdjointEmitter> {
     assert(originalValue->getFunction() == &getOriginal());
     LLVM_DEBUG(getADDebugStream() << "Adding adjoint for " << originalValue);
 #ifndef NDEBUG
-    auto origTy = remapType(originalValue->getType()).getASTType();
-    auto tangentSpace = getTangentSpace(origTy);
     // The adjoint value must be in the tangent space.
-    assert(tangentSpace && newAdjointValue.getType().getASTType()->isEqual(
-               tangentSpace->getCanonicalType()));
+    assert(newAdjointValue.getType() ==
```
Comment (dan-zheng) [inline lib/SILOptimizer/Mandatory/Differentiation.cpp]: `getRemappedTangentType` is somewhat expensive, so I kept `#ifndef NDEBUG`.

**Comment by dan-zheng** [comment]
@swift-ci Please test tensorflow

**Comment by dan-zheng** [comment]
@swift-ci Please test tensorflow

File: lib/SILOptimizer/Mandatory/Differentiation.cpp [hunk 2]
```diff
@@ -3893,11 +3900,9 @@ class AdjointEmitter final : public SILInstructionVisitor<AdjointEmitter> {
     assert(originalValue->getFunction() == &getOriginal());
     LLVM_DEBUG(getADDebugStream() << "Adding adjoint for " << originalValue);
 #ifndef NDEBUG
```
Comment (rxwei) [inline lib/SILOptimizer/Mandatory/Differentiation.cpp]: Please remove `#ifndef`/`#endif` if every debug expression is in the parentheses of an `assert`.

File: lib/SILOptimizer/Mandatory/Differentiation.cpp [hunk 3]
```diff
@@ -3863,6 +3862,14 @@ class AdjointEmitter final : public SILInstructionVisitor<AdjointEmitter> {
   void initializeAdjointValue(SILBasicBlock *origBB, SILValue originalValue,
                               AdjointValue adjointValue) {
     assert(origBB->getParent() == &getOriginal());
+    assert(originalValue->getType().isObject());
+    assert(adjointValue.getType().isObject());
+    assert(originalValue->getFunction() == &getOriginal());
+#ifndef NDEBUG
```
Comment (rxwei) [inline lib/SILOptimizer/Mandatory/Differentiation.cpp]: Same here.

**Review by rxwei — APPROVED** [review]

File: lib/SILOptimizer/Mandatory/Differentiation.cpp (hunk 1 above)
Comment (rxwei) [reply lib/SILOptimizer/Mandatory/Differentiation.cpp]: That actually won't make a difference. `assert` is being compiled away anyways.

Comment (dan-zheng) [reply lib/SILOptimizer/Mandatory/Differentiation.cpp]: Removing `#ifndef NDEBUG` based on [feedback below](https://github.com/apple/swift/pull/25264#discussion_r290892271).

File: lib/SILOptimizer/Mandatory/Differentiation.cpp (hunk 2 above)
Comment (dan-zheng) [reply lib/SILOptimizer/Mandatory/Differentiation.cpp]: I see, that makes sense.

I found two other usages of `#ifndef NDEBUG`. This usage doesn't seem quite right to me:
```
  void visit(SILInstruction *inst) {
    if (errorOccurred)
      return;

    LLVM_DEBUG(getADDebugStream() << "AdjointEmitter visited:\n[ORIG]"
               << *inst);
#ifndef NDEBUG
    auto beforeInsertion = std::prev(builder.getInsertionPoint());
#endif
    SILInstructionVisitor::visit(inst);
    LLVM_DEBUG({
      auto &s = llvm::dbgs() << "[ADJ] Emitted:\n";
      auto afterInsertion = builder.getInsertionPoint();
      for (auto it = ++beforeInsertion; it != afterInsertion; ++it)
        s << *it;
    });
  }
```

Would this be more correct? (double up `NDEBUG` and `LLVM_DEBUG`)
```diff
   void visit(SILInstruction *inst) {
     if (errorOccurred)
       return;

+#ifndef NDEBUG
     LLVM_DEBUG(getADDebugStream() << "AdjointEmitter visited:\n[ORIG]"
                << *inst);
-#ifndef NDEBUG
     auto beforeInsertion = std::prev(builder.getInsertionPoint());
 #endif
     SILInstructionVisitor::visit(inst);
+#ifndef NDEBUG
     LLVM_DEBUG({
       auto &s = llvm::dbgs() << "[ADJ] Emitted:\n";
       auto afterInsertion = builder.getInsertionPoint();
       for (auto it = ++beforeInsertion; it != afterInsertion; ++it)
         s << *it;
     });
+#endif
   }
```

Comment (rxwei) [reply lib/SILOptimizer/Mandatory/Differentiation.cpp]: No. `#ifndef NDEBUG` make code that's in the parentheses of macros that are to be compiled away compiled away. In order words, `LLVM_DEBUG` and `assert` are compiled away so you don't need `#ifndef` around them.

Comment (dan-zheng) [reply lib/SILOptimizer/Mandatory/Differentiation.cpp]: > No. #ifndef NDEBUG make code that's in the parentheses of macros that are to be compiled away compiled away.

I think I understand now, thanks!

**Comment by dan-zheng** [comment]
@swift-ci Please test tensorflow

**Comment by dan-zheng** [comment]
@swift-ci Please test tensorflow

**Comment by dan-zheng** [comment]
@swift-ci Please test tensorflow

**Pull request merged**